pkg/crypto/tinkcrypto/primitive/aead/subtle/aes_cbc.go

```diff
@@ -0,0 +1,154 @@
+/*
+Copyright SecureKey Technologies Inc. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package subtle
+
+import (
+	"bytes"
+	"crypto/aes"
+	"crypto/cipher"
+	"errors"
+	"fmt"
+
+	"github.com/google/tink/go/subtle/random"
+)
+
+// AESCBC is an implementation of AEAD interface.
+type AESCBC struct {
+	Key []byte
+}
+
+// NewAESCBC returns an AESCBC instance.
+// The key argument should be the AES key, either 16, 24 or 32 bytes to select
+// AES-128, AES-192 or AES-256.
+func NewAESCBC(key []byte) (*AESCBC, error) {
+	keySize := uint32(len(key))
+	if err := ValidateAESKeySize(keySize); err != nil {
+		return nil, fmt.Errorf("aes_cbc: NewAESCBC() %w", err)
+	}
+
+	return &AESCBC{Key: key}, nil
+}
+
+// Encrypt encrypts plaintext using AES in CTR mode.
+// The resulting ciphertext consists of two parts:
+// (1) the IV used for encryption and (2) the actual ciphertext.
+func (a *AESCBC) Encrypt(plaintext []byte) ([]byte, error) {
+	plainTextSize := len(plaintext)
+	if plainTextSize > maxInt-AESCBCIVSize {
+		return nil, errors.New("aes_cbc: plaintext too long")
+	}
+
+	iv := a.newIV()
+
+	cbc, err := newCipher(a.Key, iv, false)
+	if err != nil {
+		return nil, fmt.Errorf("aes_cbc: Encrypt() %w", err)
+	}
+
+	ciphertext := make([]byte, AESCBCIVSize+plainTextSize)
+	if n := copy(ciphertext, iv); n != AESCBCIVSize {
+		return nil, fmt.Errorf("aes_cbc: failed to copy IV (copied %d/%d bytes)", n, AESCBCIVSize)
+	}
+
+	if n := copy(ciphertext[AESCBCIVSize:], plaintext); n != plainTextSize {
+		return nil, fmt.Errorf("aes_cbc: failed to copy plaintext (copied %d/%d bytes)", n, plainTextSize)
+	}
+
+	ciphertext = Pad(ciphertext, plainTextSize, cbc.BlockSize())
+
+	cbc.CryptBlocks(ciphertext[AESCBCIVSize:], ciphertext[AESCBCIVSize:])
+
+	return ciphertext, nil
+}
+
+// Decrypt decrypts ciphertext.
+func (a *AESCBC) Decrypt(ciphertext []byte) ([]byte, error) {
+	ciphertextSize := len(ciphertext)
+	if ciphertextSize < AESCBCIVSize {
+		return nil, errors.New("aes_cbc: ciphertext too short")
+	}
+
+	iv := ciphertext[:AESCBCIVSize]
+
+	cbc, err := newCipher(a.Key, iv, true)
+	if err != nil {
+		return nil, fmt.Errorf("aes_cbc: Decrypt() %w", err)
+	}
+
+	blockSize := cbc.BlockSize()
+
+	if len(ciphertext[AESCBCIVSize:])%blockSize > 0 {
+		return nil, errors.New("aes_cbc: invalid ciphertext padding")
+	}
+
+	plaintext := make([]byte, ciphertextSize-AESCBCIVSize)
+	cbc.CryptBlocks(plaintext, ciphertext[AESCBCIVSize:])
+
+	if len(plaintext) == 0 {
+		return plaintext, nil
+	}
+
+	// unpad plaintext if not block size.
+	last := plaintext[len(plaintext)-1]
+	count := int(last)
+
+	if count == 0 || count > blockSize || count > len(plaintext) {
```

Comment:
This looks wrong to me, because it'd reject any 16-byte plaintext that ends with a byte that is 0 or 17 or higher.

You don't want to make padding optional. Even if the plaintext length is already a multiple of the block size, add a dummy block consisting of 16 bytes of value 15.

This also helps make sure the ciphertext of the empty plaintext is indistinguishable.

Reply:
thank you @thaidn, it makes sense, but why this last block must contain bytes of value 15? 

What if the plaintext needs padding, shouldn't this extra block contain bytes of the padding value which is the padding size, eg if plaintext is missing 6 bytes to the block, then pad 6 bytes of value 6, and add a new block of 16 bytes of value 6, not 15? Or are you saying padding should always be hard coded with value 15 for both the dummy block and the padding bytes (for plaintext length not multiple of the block size)?

Reply:
15 is a typo, it should be 16.

You want to implement: https://en.wikipedia.org/wiki/Padding_(cryptography)#PKCS#5_and_PKCS#7.

The value of each added byte is the number of bytes that are added, i.e. N bytes, each of value N are added. Let's say the plaintext size is 5. You would need to add 9 bytes, each of value 9. If the plaintext size is 16, you'd need to add a whole block of padding, meaning 16 bytes, each of value 16.

Reply:
So adding a full dummy block is only added when plaintext's size is a multiple of block size. This doesn't match the NIST examples though, since plaintext size is a multiple of block size in these examples. See CBC test vectors on page 27 of the [NIST document 800-3a](https://nvlpubs.nist.gov/nistpubs/Legacy/SP/nistspecialpublication800-38a.pdf), plaintext are not padded with a dummy buffer.

what to do in this case? 

In the meantime I will push a pr that adds a dummy block for plaintext messages with size being a multiple of block size.